### PR TITLE
WIP Objection

### PR DESCRIPTION
**Pull request by eopo**
# Description
Added [http://github.com/Vincit/objection.js](Vincit/objection.js) as Model layer.

## Type of change

Please delete options that are not relevant.

- [x] Breaking Change (Modified API)
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?
Live instance fed by MessageRepeat running from this branch - With no problems up to now. Additional testing welcome, since no edge cases were tested as of now.

# Checklist:

- [X] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated changelog.md with changes made

### CODE REVIEW

**Comment by s3m1s0n1c** [comment]
Ooo Interesting... Hope you get it working with out the issues @DanrwAU had with performance

**Comment by marshyonline** [comment]
Going to need to leave this one with me for a few days - need to spin up an instance and push traffic into it

**Comment by DanrwAU** [comment]
@pagermon/collaborators I think this one should probably bump us a minor version, it's another major DB driver change... thoughts?

I'd suggest this one gets PR'd into v0.4.0. 

@eopo i'll spin this up in azure this evening and push all my prod traffic, see if i can break it, the major thing i found when implementing Knex was ensuring Searching etc was tested with all the possible combinations of ApiSecurity and HideCapcode, along with PDWMode etc. 
Prime candidate for someone to write some tests 😆

**Comment by DanrwAU** [comment]
Searching is currently crashing 😢 

```
[db]  2019-11-03 18:45:48  debug : {"method":"select","options":{},"timeout":false,"cancelOnTimeout":false,"bindings":["0","test",20],"__knexQueryUid":"3f7236db-4153-4508-9cd6-b6ff9371e415","sql":"select `messages`.*, `alias`.`alias`, `alias`.`aliasMatch`, `alias`.`icon`, `alias`.`agency`, `alias`.`color`, `alias`.`ignore` from `messages` left join (select `alias`, `agency`, `icon`, `color`, `ignore`, `pluginconf`, `id` as `aliasMatch`, `capcodes`.`id` from `capcodes`) as `alias` on `alias`.`id` = `messages`.`alias_id` where `alias`.`ignore` = ? and MATCH(message, address, source) AGAINST (? IN BOOLEAN MODE) order by `timestamp` DESC limit ?"}

You have triggered an unhandledRejection, you may have forgotten to catch a Promise rejection:
Error: SQLITE_ERROR: near "AGAINST": syntax error

```

**Comment by DanrwAU** [comment]
Posting messages is also returning a 500 error on insert

**Comment by eopo** [comment]
> [db]  2019-11-03 18:45:48  debug : {"method":"select","options":{},"timeout":false,"cancelOnTimeout":false,"bindings":["0","test",20],"__knexQueryUid":"3f7236db-4153-4508-9cd6-b6ff9371e415","sql":"select `messages`.*, `alias`.`alias`, `alias`.`aliasMatch`, `alias`.`icon`, `alias`.`agency`, `alias`.`color`, `alias`.`ignore` from `messages` left join (select `alias`, `agency`, `icon`, `color`, `ignore`, `pluginconf`, `id` as `aliasMatch`, `capcodes`.`id` from `capcodes`) as `alias` on `alias`.`id` = `messages`.`alias_id` where `alias`.`ignore` = ? and MATCH(message, address, source) AGAINST (? IN BOOLEAN MODE) order by `timestamp` DESC limit ?"}
> 
> You have triggered an unhandledRejection, you may have forgotten to catch a Promise rejection:
> Error: SQLITE_ERROR: near "AGAINST": syntax error
> ```

Ahh shoot. Forgot adapting it for sqlite. MySQL works fine for me, but SQLite has this f**** different syntax.

**Comment by DanrwAU** [comment]
Yup 😉

Welcome to my nightmare 🤣

**Comment by eopo** [comment]
Need more info on that. Works fine on my MariaDB instance, but that doesn't mean a thing ...

**Comment by eopo** [comment]
Could you please try this one? I don't have a sqlite instance running atm, you might be quicker with testing ;)

**Comment by DanrwAU** [comment]
> Could you please try this one? I don't have a sqlite instance running atm, you might be quicker with testing ;)

It no longer crashes, but doesn't work either.

None of these results should be matching
image

**Comment by DanrwAU** [comment]
> Need more info on that. Works fine on my MariaDB instance, but that doesn't mean a thing ...

This is the logs from trying to post a message to the api 

```
[pmon]  2019-11-04 00:18:43  debug : beforeMessage start
[pmon]  2019-11-04 00:18:43  debug : ======================
[pmon]  2019-11-04 00:18:43  debug : trigger: message scope: before
[pmon]  2019-11-04 00:18:43  debug : ======================
[pmon]  2019-11-04 00:18:43  debug : data object
[pmon]  2019-11-04 00:18:43  debug : { address: '1908434',
  message: 'HbPlease provide ESTA Dispatch an update of your status2',
  source: 'Test',
  pluginData: {} }
[pmon]  2019-11-04 00:18:43  debug : plugins object
[pmon]  2019-11-04 00:18:43  debug : {}
[pmon]  2019-11-04 00:18:43  debug : ======================
[pmon]  2019-11-04 00:18:43  debug : Before Plugin return: undefined
[pmon]  2019-11-04 00:18:43  debug : beforeMessage done
[db]  2019-11-04 00:18:43  debug : {"method":"select","options":{},"timeout":false,"cancelOnTimeout":false,"bindings":[-59,"HbPlease provide ESTA Dispatch an update of your status2","1908434",10],"__knexQueryUid":"1934093f-2feb-40a0-834c-7a1064d61e56","sql":"select `messages`.* from `messages` where `timestamp` > ? and `message` = ? and `address` = ? order by `id` DESC limit ?"}
[db]  2019-11-04 00:18:43  debug : {"method":"select","options":{},"timeout":false,"cancelOnTimeout":false,"bindings":["1908434"],"__knexQueryUid":"fba635e1-cd19-430d-b811-cf1a6a3881ba","sql":"select `capcodes`.* from `capcodes` where ? LIKE address order by REPLACE(address, '_', '%') DESC"}
[db]  2019-11-04 00:18:43  debug : {"method":"insert","options":{},"timeout":false,"cancelOnTimeout":false,"bindings":["1908434","HbPlease provide ESTA Dispatch an update of your status2","Test",1],"__knexQueryUid":"cf04160f-4000-4406-9c4b-3ff50010aaaf","sql":"insert into `messages` (`address`, `message`, `source`, `timestamp`) values (?, ?, ?, ?)"}
[db]  2019-11-04 00:18:43  debug : {"method":"select","options":{},"timeout":false,"cancelOnTimeout":false,"bindings":[339534],"__knexQueryUid":"ff0e8810-92b9-44f2-8f4f-cadf91cf5c3b","sql":"select `messages`.* from `messages` where `messages`.`id` in (?)"}
[db]  2019-11-04 00:18:43  debug : {"method":"select","options":{},"timeout":false,"cancelOnTimeout":false,"bindings":[339534],"__knexQueryUid":"55a0fef8-3b7c-4846-ae2e-74b97a451516","sql":"select `messages`.*, `alias`.* from `messages` inner join `capcodes` as `alias` on `alias`.`id` = `messages`.`alias_id` where `messages`.`id` = ?"}
::1 - - [03/Nov/2019:13:18:43 +0000] "POST /api/messages HTTP/1.1" 500 2 "-" "PostmanRuntime/7.19.0"

```

This is what is being sent 
```
{ 
	"address":"1908434",
	"message":"HbPlease provide ESTA Dispatch an update of your status2",
	"source":"Test"
}
```

Which i use to test current versions just fine 😁

**Comment by eopo** [comment]
Have fun testing. Works fine with SQLITE for me atm ;)

**Comment by marshyonline** [comment]
Seems like I've found 1 bug so far
SQLlight as the DB backend
http://localhost:3000/api/messages/%MESSAGEID%

```[pmon]  2019-11-04 12:06:21  debug : calling next 69
[db]  2019-11-04 12:06:21  debug : {"method":"select","options":{},"timeout":false,"cancelOnTimeout":false,"bindings":["7"],"__knexQueryUid":"1dd03cac-3a58-43b9-8804-133a50fdda33","sql":"select `messages`.* from `messages` where `messages`.`id` = ? order by `timestamp` DESC"}
[db]  2019-11-04 12:06:21  debug : {"method":"select","options":{},"timeout":false,"cancelOnTimeout":false,"bindings":[1,0],"__knexQueryUid":"22e9b668-e00a-4344-bb23-2fd8b6ef372e","sql":"select `capcodes`.`id`, `alias`, `agency`, `icon`, `color`, `ignore`, `id` as `aliasMatch` from `capcodes` where `capcodes`.`id` in (?) and `ignore` = ?"}
sql: 9.926ms
(node:441903) Warning: Label 'send' already exists for console.time()
(node:441903) UnhandledPromiseRejectionWarning: TypeError: queryResult.catch is not a function
    at /home/nathanial/git/pagermon-eopo/server/routes/api.js:390:24
(node:441903) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 3)
::1 - - [04/Nov/2019:01:06:21 +0000] "GET /api/messages/7 HTTP/1.1" 304 - "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.120 Safari/537.36"```

**Comment by DanrwAU** [comment]
> Seems like I've found 1 bug so far
> SQLlight as the DB backend
> http://localhost:3000/api/messages/%MESSAGEID%
> 
> ```
> [db]  2019-11-04 12:06:21  debug : {"method":"select","options":{},"timeout":false,"cancelOnTimeout":false,"bindings":["7"],"__knexQueryUid":"1dd03cac-3a58-43b9-8804-133a50fdda33","sql":"select `messages`.* from `messages` where `messages`.`id` = ? order by `timestamp` DESC"}
> [db]  2019-11-04 12:06:21  debug : {"method":"select","options":{},"timeout":false,"cancelOnTimeout":false,"bindings":[1,0],"__knexQueryUid":"22e9b668-e00a-4344-bb23-2fd8b6ef372e","sql":"select `capcodes`.`id`, `alias`, `agency`, `icon`, `color`, `ignore`, `id` as `aliasMatch` from `capcodes` where `capcodes`.`id` in (?) and `ignore` = ?"}
> sql: 9.926ms
> (node:441903) Warning: Label 'send' already exists for console.time()
> (node:441903) UnhandledPromiseRejectionWarning: TypeError: queryResult.catch is not a function
>     at /home/nathanial/git/pagermon-eopo/server/routes/api.js:390:24
> (node:441903) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 3)
> ::1 - - [04/Nov/2019:01:06:21 +0000] "GET /api/messages/7 HTTP/1.1" 304 - "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.120 Safari/537.36"```
> ```

Just a note this one isn't happening to me, not sure what the dealio is there...

**Comment by marshyonline** [comment]
Also - I agree, This should be the next major release. 
Repo split should be before the major tho

**Comment by eopo** [comment]
As long as it’s an unhandled rejection it’s bad anyways. I’ll have a look this afternoon (That is ~10 hours from now for me)

**Comment by eopo** [comment]
> (node:441903) UnhandledPromiseRejectionWarning: TypeError: queryResult.catch is not a function
>     at /home/nathanial/git/pagermon-eopo/server/routes/api.js:390:24
> (node:441903) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 3)

Fixed. Thee shall not try to catch an already resolved promise. That throws an error. If you throw an error in an async function and don't catch this one, the async will reject. Don't be like me. Be smart.

File: server/routes/api.js [hunk 1]
```diff
@@ -369,26 +369,31 @@ router.get('/messages/:id', async (req, res, next) => {
...
+            .modify(builder => {
+                if (HideCapcode && !req.isAuthenticated())
+                    builder.omit(Message, ['address']);
+            });
+
+        console.timeEnd('sql');
+        console.time('send');
+
+        if (!queryResult || (queryResult.ignore || (pdwMode && !queryResult.alias)))
+            res.status(200).json({});
```
Comment (marshyonline) [inline server/routes/api.js]: Think we are missing a time end here

Comment (eopo) [reply server/routes/api.js]: Nope, comes after the if/else to include empty result handling AND regular result handling.

Comment (eopo) [reply server/routes/api.js]: Line 390 if you ask GitHub; 394 if you ask my IDE. I'm kinda worried where the 3 lines are that I'm not aware of ...

**Comment by eopo** [comment]
Additional idea: Adding a property to plugin config that sets a required plugin API version. Plugins that can handle the newer format (With alias data in a sub-object) get it, plugins that don't define anything get the old format with the alias properties moved one level up.
This way we could implement #338 without having to look for backwards compatibility

**Comment by DanrwAU** [comment]
I forget where did we get up to with this one? 

@davidmckenzie s changes in #350 conflict with this, so we'll need to sort out which gets merged first.

**Comment by eopo** [comment]
Theoretically it would be better if we merged #337 first and #350 second.
Since objection directly uses the knex api it should be no problem to move the modifications in Oracle into the "refactored" queries in Objection. Some parts are heavily modified in both branches, I guess we have to look which version makes more sense for each line manually. It's just a lot of work.

**Comment by DanrwAU** [comment]
> Theoretically it would be better if we merged #337 first and #350 second.
> Since objection directly uses the knex api it should be no problem to move the modifications in Oracle into the "refactored" queries in Objection. Some parts are heavily modified in both branches, I guess we have to look which version makes more sense for each line manually. It's just a lot of work.

Agreed, i think it's less work if we do #350 first as most of it stays the same so it's just adapting the new stuff to objection. 

Doing #337 first will require a merge commit for #350 and be a mess to fix i think.

**Pull request closed without merging**